nntrainer/tensor/cl_operations/attention_kernels_fp16.cpp

```diff
@@ -0,0 +1,279 @@
+                   unsigned int batch, unsigned int channel,
+                   unsigned int height, unsigned int width, unsigned int dim,
+                   unsigned int from, unsigned int max_timestep,
+                   unsigned int in_size, unsigned int out_size,
+                   RunLayerContext &context) {
+
+  bool result = false;
+  do {
+    result = context.clCreateKernel(rotary_emb_cl_kernel_fp16,
+                                    context.LayerKernel::ROTARY_EMB_FP16,
+                                    kernel_rotary_emb_fp16);
+    if (!result) {
+      printf("Failed to create kernel for rotary_emb_cl\n");
+      break;
+    }
+
+    unsigned int cos_dim = cos_.size();
+    unsigned int sin_dim = sin_.size();
+    unsigned int freqs_cos_dim = freqs_cos.size();
+    unsigned int freqs_sin_dim = freqs_sin.size();
+
+    size_t dim1_size = sizeof(cl_half) * in_size;
+    size_t dim2_size = sizeof(cl_half) * out_size;
+    size_t dim3_size = sizeof(float) * cos_dim;
+    size_t dim4_size = sizeof(float) * sin_dim;
+    size_t dim5_size = sizeof(float) * freqs_cos_dim * dim;
+    size_t dim6_size = sizeof(float) * freqs_sin_dim * dim;
+
+    opencl::Buffer inputA(context.context_inst_, dim1_size, true, nullptr);
+
+    opencl::Buffer inOutRes(context.context_inst_, dim2_size, true, nullptr);
+
+    opencl::Buffer cosBuf(context.context_inst_, dim3_size, true, nullptr);
+
+    opencl::Buffer sinBuf(context.context_inst_, dim4_size, true, nullptr);
+
+    opencl::Buffer freqs_cosBuf(context.context_inst_, dim5_size, true,
+                                nullptr);
+
+    opencl::Buffer freqs_sinBuf(context.context_inst_, dim6_size, true,
+                                nullptr);
+
+    std::vector<float> freqs_cos_flat;
+    std::vector<float> freqs_sin_flat;
+    for (const auto &row : freqs_cos) {
+      freqs_cos_flat.insert(freqs_cos_flat.end(), row.begin(), row.end());
+    }
+    for (const auto &row : freqs_sin) {
+      freqs_sin_flat.insert(freqs_sin_flat.end(), row.begin(), row.end());
+    }
+
+    result = inputA.WriteData(context.command_queue_inst_, in);
+    if (!result) {
+      printf("Failed to write input data\n");
+      break;
+    }
+
+    result = inOutRes.WriteData(context.command_queue_inst_, out);
+    if (!result) {
+      printf("Failed to write output data\n");
+      break;
+    }
+
+    result = freqs_cosBuf.WriteData(context.command_queue_inst_,
+                                    freqs_cos_flat.data());
+    if (!result) {
+      printf("Failed to write cos data\n");
+      break;
+    }
+
+    result = freqs_sinBuf.WriteData(context.command_queue_inst_,
+                                    freqs_sin_flat.data());
+    if (!result) {
+      printf("Failed to write sin data\n");
+      break;
+    }
+
+    result = cosBuf.WriteData(context.command_queue_inst_, cos_.data());
+    if (!result) {
+      printf("Failed to write cos data\n");
```

Comment:
Like the "Failed to set ~" log, how about making the freqs_cosBuf and cosBuf logs different in the "Failed to write ~" log?

Reply:
Yeah sure, Made the necessary changes in the latest commit. Thanks for pointing it out.